server/utils/cloudFunctions/functions/countAdvertisersCloudFunctions/countAdvertisersHelpers.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Description: Helper functions for count advertisers cloud functions.
+ * Author: Robert Marcus
+ * Date: July 14, 2020
+ */
+
+const { assert } = require('../test/testConfig');
+const { firestoreFieldValue } = require('./countAdvertisersConfig');
+
+/**
+ * Assert `advertiser` is a string
+ * @param {string} advertiser as derived from snapshot
+ */
+function checkAdvertiser(advertiser) {
+  assert.typeOf(advertiser, 'string');
+}
+
+/**
+ * Assert `startDate` is a string and a valid iso date string.
+ * If `startDate` is not a valid iso string, `isoStartDate` will == NaN.
+ * @param {string} startDate as derived from snapshot
+ * @param {Object} isoStartDate Date(startDate) output, json string
+ */
+function checkStartDate(startDate, isoStartDate) {
+  assert.typeOf(startDate, 'string');
+  assert.ok(isoStartDate.getFullYear()); 
+}
+
+/**
+ * Returns a filepath reference to a doc containing the numberOfAds an 
+ * advertiser released in a particular year. Will link to whatever year 
+ * the ad in the `snapshot` is from. 
+ * If `advertiser` or `startDate` field is not well-typed/formed, 
+ * an assertion error will be thrown.     
+ * @param {Object} snapshot a json string
+ * @param {Object} collection reference to a particular firestore collection
+ * @returns {Object} 
+ */
+function getAdvertiserCountReference(snapshot, collection) {
+  const data = snapshot.data();
+  const advertiser = data.advertiser;
+  const startDate = data.startDate;
+  const isoStartDate = new Date(startDate);
+
+  checkAdvertiser(advertiser);
+  checkStartDate(startDate, isoStartDate);
+
+  const startYear = isoStartDate.getFullYear().toString(); 
+
+  const advertiserCountReference = 
+      collection.doc(startYear)
+                .collection("advertisers")
+                .doc(advertiser);
+
+  return advertiserCountReference;
+}
+
+/**
+ * Decrements an advertiser's annual ad count by one for an ad in a given year.
+ * Event triggered on delete. If the advertiser's count document does not 
+ * exist in the relevant collection, a 500 error will be thrown.
+ * If `advertiser` or `startDate` field is not well-typed/formed, 
+ * an assertion error will be thrown. 
+ * @param {Object} snapshot a json string
+ * @param {Object} collection reference to a particular firestore collection
+ * @returns {!Promise} 
+ */
+function decrementAdvertiserCount(snapshot, collection) {
+  const advertiserCountReference = 
+      getAdvertiserCountReference(snapshot, collection);
+
+  return advertiserCountReference.get().then(function(doc) {
+    if (doc.exists) {
+
+      // Only decrement if the numberOfAds is positive.
+      const change = doc.data().numberOfAds > 0 ? -1 : 0;
```

Comment:
Why would we ever decrement when the advertiser already believes it has zero ads?  This would indicate that your data is already in an invalid state and you should at least log a warning.

Reply:
Will rewrite this to include a log.